windows/src/desktop/kmshell/main/Keyman.System.RemoteUpdateCheck.pas

```diff
@@ -0,0 +1,386 @@
+unit Keyman.System.RemoteUpdateCheck;  // I3306
+
+interface
+
+uses
+  System.Classes,
+  System.SysUtils,
+  KeymanPaths,
+  httpuploader,
+  Keyman.System.UpdateCheckResponse,
+  OnlineUpdateCheck;
+
+type
+  ERemoteUpdateCheck = class(Exception);
+
+  TRemoteUpdateCheckResult = (wucUnknown, wucSuccess, wucNoUpdates, wucFailure, wucOffline);
+
+  TRemoteUpdateCheckDownloadParams = record
+    TotalSize: Integer;
+    TotalDownloads: Integer;
+    StartPosition: Integer;
+  end;
+
+  TRemoteUpdateCheck = class
+  private
+    FForce: Boolean;
+    FRemoteResult: TRemoteUpdateCheckResult;
+
+    FErrorMessage: string;
+
+    FShowErrors: Boolean;
+    FDownload: TRemoteUpdateCheckDownloadParams;
+    FCheckOnly: Boolean;
+
+    function DownloadUpdates(Params: TUpdateCheckResponse) : Boolean;
+    procedure DoDownloadUpdates(SavePath: string; Params: TUpdateCheckResponse;  var Result: Boolean);
+    function DoRun: TRemoteUpdateCheckResult;
+  public
+
+    constructor Create(AForce : Boolean; ACheckOnly: Boolean = False);
+    destructor Destroy; override;
+    function Run: TRemoteUpdateCheckResult;
+    property ShowErrors: Boolean read FShowErrors write FShowErrors;
+  end;
+
+procedure LogMessage(LogMessage: string);
+
+implementation
+
+uses
+  System.WideStrUtils,
+  Winapi.Windows,
+  Winapi.WinINet,
+
+  GlobalProxySettings,
+  KLog,
+  keymanapi_TLB,
+  KeymanVersion,
+  Keyman.System.UpdateCheckStorage,
+  kmint,
+  ErrorControlledRegistry,
+  RegistryKeys,
+  Upload_Settings,
+
+  OnlineUpdateCheckMessages;
+
+{ TRemoteUpdateCheck }
+
+constructor TRemoteUpdateCheck.Create(AForce, ACheckOnly: Boolean);
+begin
+  inherited Create;
+
+  FShowErrors := True;
+  FRemoteResult := wucUnknown;
+
+  FForce := AForce;
+  FCheckOnly := ACheckOnly;
+
+  KL.Log('TRemoteUpdateCheck.Create');
+end;
+
+destructor TRemoteUpdateCheck.Destroy;
+begin
+  if (FErrorMessage <> '') and FShowErrors then
+    LogMessage(FErrorMessage);
+
+  KL.Log('TRemoteUpdateCheck.Destroy: FErrorMessage = '+FErrorMessage);
+  KL.Log('TRemoteUpdateCheck.Destroy: FRemoteResult = '+IntToStr(Ord(FRemoteResult)));
+
+  inherited Destroy;
+end;
+
+function TRemoteUpdateCheck.Run: TRemoteUpdateCheckResult;
+begin
+  Result := DoRun;
+
+  if Result in [ wucSuccess] then
+  begin
+    kmcom.Keyboards.Refresh;
+    kmcom.Keyboards.Apply;
+    kmcom.Packages.Refresh;
+  end;
+
+  FRemoteResult := Result;
+end;
+
+
+procedure TRemoteUpdateCheck.DoDownloadUpdates(SavePath: string; Params: TUpdateCheckResponse; var Result: Boolean);
+var
+  i, downloadCount: Integer;
+
+    function DownloadFile(const url, savepath: string): Boolean;
+    begin
+      with THttpUploader.Create(nil) do
+      try
+        Proxy.Server := GetProxySettings.Server;
+        Proxy.Port := GetProxySettings.Port;
+        Proxy.Username := GetProxySettings.Username;
+        Proxy.Password := GetProxySettings.Password;
+        Request.Agent := API_UserAgent;
+
+        Request.SetURL(url);
+        Upload;
```

Comment:
Strange that you have to call `Upload` and use a `THttpUploader` to download a file...

Reply:
`THttpUploader` was originally only for upload. In the Bad Old Days renaming things was hard. We could rename it now, but remember that all the Delphi code is essentially in techdebt state so I am not keen on investing in refactoring it much.